### PR TITLE
Implicit type references

### PR DESCRIPTION
**Pull request by RyanCavanaugh**
Refactors implicit type references based on discussion.

This required a bunch of work in the harness and we should rethink how the compiler runner creates unit names, as it's currently totally wacky.

____

*Edit by @DanielRosenwasser: Fixes #9012*

### CODE REVIEW

File: src/compiler/program.ts [hunk 1]
```diff
@@ -972,21 +959,93 @@ namespace ts {
...
+
+    /**
+      * Computes the types folder, a.k.a. the primary types lookup location.
+      * This is commonly './types' relative to whatever the compilation root is.
+      */
+    function getTypesFolderLocation(options: CompilerOptions, host: CompilerHost, inferredRoot: string): Path {
+        // Openish question: When tsconfig specifies a rootDir *and* a relative typesRoot, is the typesRoot
+        // relative to the tsconfig file path, or the rootDir path?
+
+        // If typesRoot came from tsconfig, then it's relative the config file path.
```
Comment (DanielRosenwasser) [inline src/compiler/program.ts]: to the

File: src/compiler/program.ts [hunk 2]
```diff
@@ -972,21 +959,93 @@ namespace ts {
...
 
-    export function getDefaultTypeDirectiveNames(options: CompilerOptions, rootFiles: string[], host: CompilerHost): string[] {
+
+    /**
+      * Computes the types folder, a.k.a. the primary types lookup location.
+      * This is commonly './types' relative to whatever the compilation root is.
+      */
+    function getTypesFolderLocation(options: CompilerOptions, host: CompilerHost, inferredRoot: string): Path {
+        // Openish question: When tsconfig specifies a rootDir *and* a relative typesRoot, is the typesRoot
+        // relative to the tsconfig file path, or the rootDir path?
```
Comment (DanielRosenwasser) [inline src/compiler/program.ts]: I started out thinking that it should be relative to the `rootDir`, but I think that if someone has an *explicit* relative `typesRoot`, then that should be relative to the `tsconfig.json`.

A user should be able to open up a tsconfig and know the `typesRoot` without doing any extra calculations.

Comment (DanielRosenwasser) [inline src/compiler/program.ts]: @mhegazy @vladima what say ye?

File: src/compiler/program.ts [hunk 3]
```diff
@@ -972,21 +959,93 @@ namespace ts {
...
+
+        // If typesRoot came from tsconfig, then it's relative the config file path.
+        // Otherwise, it's relative to the current folder.
+        if (options.typesRoot) {
+            // true case is e.g. tsconfig { typesRoot: './foo' }, so types is [tsconfigPath]/foo
+            // false case is e.g. tsc a.ts --typesRoot foo, so types is [cwd]/foo
```
Comment (DanielRosenwasser) [inline src/compiler/program.ts]: Be explicit that this is also the empty string case.

Comment (DanielRosenwasser) [inline src/compiler/program.ts]: Why do you get a false case if you specify the option from `tsc`?

Comment (DanielRosenwasser) [inline src/compiler/program.ts]: What if you had written `--typesRoot ./foo`?

Comment (RyanCavanaugh) [inline src/compiler/program.ts]: We don't even have that information at this point; I'm inclined to say if you use `--p` *and* specify extra commandline values, the commandline values are still validly interpreted as being relative to the project file path

Comment (DanielRosenwasser) [inline src/compiler/program.ts]: I guess I should clarify.

I'm asking why the `options.typesRoot` false case is triggered when it's specified on the command line. It looks like you're specifying `typesRoot` in either case in the above comment.

Comment (RyanCavanaugh) [inline src/compiler/program.ts]: This comment refers to the two branches of the `?` operator on the following line, not the preceding line

**Comment by RyanCavanaugh** [comment]
Closing while I rework based on feedback from today's design meeting

**Pull request closed without merging**